plugin/plugins/filesystem/filesystem.go

```diff
@@ -0,0 +1,124 @@
+package filesystem
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/djdv/p9/p9"
+	plugin "github.com/ipfs/go-ipfs/plugin"
+	fsnodes "github.com/ipfs/go-ipfs/plugin/plugins/filesystem/nodes"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+var (
+	_ plugin.PluginDaemon = (*FileSystemPlugin)(nil) // impl check
+
+	// Plugins is an exported list of plugins that will be loaded by go-ipfs.
+	Plugins = []plugin.Plugin{
+		&FileSystemPlugin{}, //TODO: individually name implementations: &P9{}
+	}
+
+	logger logging.EventLogger
+)
+
+func init() {
+	logger = logging.Logger("plugin/filesystem")
+}
+
+type FileSystemPlugin struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	addr      multiaddr.Multiaddr
+	listener  manet.Listener
+	errorChan chan error
+}
+
+func (*FileSystemPlugin) Name() string {
+	return PluginName
+}
+
+func (*FileSystemPlugin) Version() string {
+	return PluginVersion
+}
+
+func (fs *FileSystemPlugin) Init(env *plugin.Environment) error {
+	logger.Info("Initializing 9P resource server...")
+	if !filepath.IsAbs(env.Repo) {
+		absRepo, err := filepath.Abs(env.Repo)
+		if err != nil {
+			return err
+		}
+		env.Repo = absRepo
+	}
+
+	cfg := &Config{}
+	// config not being set is okay and will load defalts
+	// config being set with malformed data is not okay and will instruct the daemon to halt-and-catch-fire
+	rawConf, ok := (env.Config).(json.RawMessage)
```

Comment:
Re: https://github.com/ipfs/go-ipfs/pull/6612#discussion_r323739035

I meant the first one. That is, we could change the type of `env.Config` to `json.RawMessage`. _This_ change won't work as `env.Config` will _never_ be `json.RawMessage`.

Let's just revert this and leave it as-is. To convert from the `interface{}` config to a struct, we can use https://github.com/mitchellh/mapstructure.

Reply:
4717f7367e40b7ca76be82684d7f04010c8c1b7b + https://github.com/ipfs/go-ipfs/commit/21b2d2cf1b13db5076405173d02a803f4a8c241f#diff-8d243ab8514fa1b77d41df37a81641c9R50